MayssaraOmar/assignment3/Word search/src/Word search.cpp

```diff
@@ -2,47 +2,54 @@
 #include "word search.h"
 using namespace std;
 
-void get_words_from_grid(vector<vector<char> > &grid, int row, int col,
-		string word, map<pair<int, int>, bool> &vis,
-		map<pair<pair<int, int>, string>, bool> &memo, dictionary &dic,
-		unordered_set<string> &words) {
+// internal function
+// cache maps a pair of a (cell position (row, col)) and a string (prefix) to a boolean to ensure that (get_words_from_grid) method doesn't run for the same inputs more than once using memoization technique
+static void get_words_from_grid(const vector<vector<char> > &grid, int row, int col,
+		string prefix, map<pair<int, int>, bool>* visited,
+		map<pair<pair<int, int>, string>, bool>* cache, const dictionary &dic,
+		unordered_set<string>* words) {
+
+	// boundary check
 	if (row < 0 || row >= grid.size() || col < 0 || col >= grid[row].size()) {
 		return;
 	}
-	if (!dic.isPrefix(word + grid[row][col])
-			|| memo[make_pair(make_pair(row, col), word)]) {
+	// result has been already computed for these inputs
+	if((*cache)[make_pair(make_pair(row, col), prefix)]){
 		return;
 	}
-
-	if (dic.isWord(word + grid[row][col])) {
-		words.insert(word + grid[row][col]);
+	// word exits in dictionary
+	if (dic.isWord(prefix + grid[row][col])) {
+			words->insert(prefix + grid[row][col]);
```

Comment:
You are doing `prefix + grid[row][col]` in several places, may be easier to do `prefix += grid[row][col];` before this if